File: address-geocode-international-python/REST/place_search_rest.py
'''
Service Objects - AGI PlaceSearch Client

This module provides the place_search function to search and geocode international places
using Service Objects' AGI PlaceSearch REST API. It supports trial/live environments,
fallback handling, and extended parameter support.

Functions:
    place_search(single_line: str,
                address1: str,
                address2: str,
                address3: str,
                address4: str,
                address5: str,
                locality: str,
                administrative_area: str,
                postal_code: str,
                country: str,
                boundaries: str,
                max_results: int,
                search_type: str,
                extras: str,
                license_key: str,
                is_live: bool) -> dict:
'''

import requests

# Endpoint URLs for AGI PlaceSearch
PRIMARY_URL = "https://sws.serviceobjects.com/AGI/api.svc/json/PlaceSearch"
BACKUP_URL = "https://swsbackup.serviceobjects.com/AGI/api.svc/PlaceSearch"
TRIAL_URL = "https://trial.serviceobjects.com/AGI/api.svc/json/PlaceSearch"

def place_search(single_line: str,
                address1: str,
                address2: str,
                address3: str,
                address4: str,
                address5: str,
                locality: str,
                administrative_area: str,
                postal_code: str,
                country: str,
                boundaries: str,
                max_results: int,
                search_type: str,
                extras: str,
                license_key: str,
                is_live: bool) -> dict:
    """
    Calls the AGI PlaceSearch API and returns the parsed response.

    Parameters:
        single_line (str): Full address or place name (takes priority).
        address1–5 (str): Optional multiline address input fields.
        locality (str): City or locality.
        administrative_area (str): State, province, or region.
        postal_code (str): ZIP or postal code.
        country (str): ISO-2 country code.
        boundaries (str): Optional boundary filter.
        max_results (int): Number of results to return.
        search_type (str): Type of search: Address, Locality, PostalCode, etc.
        extras (str): Optional additional flags.
        license_key (str): AGI API key.
        is_live (bool): True for live endpoint; False for trial.

    Returns:
        dict: Parsed JSON response from the API.

    Raises:
        RuntimeError: If the API returns an error payload.
        requests.RequestException: On network/HTTP failures (trial mode).
    """

    # Prepare query parameters for AGI API
    params = {
        "SingleLine": single_line,
        "Address1": address1,
        "Address2": address2,
        "Address3": address3,
        "Address4": address4,
        "Address5": address5,
        "Locality": locality,
        "AdministrativeArea": administrative_area,
        "PostalCode": postal_code,
        "Country": country,
        "Boundaries": boundaries,
        "MaxResults": max_results,
        "SearchType": search_type,
        "Extras": extras,
        "LicenseKey": license_key
    }

    # Select the base URL: production vs trial
    url = PRIMARY_URL if is_live else TRIAL_URL

    # Attempt primary (or trial) endpoint first
    try:
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        data = response.json()

        # If API returned an error in JSON payload, trigger fallback
        error = getattr(response, 'Error', None)
        if not (error is None or getattr(error, 'TypeCode', None) != "3"):
            if is_live:
                # Try backup URL when live
                response = requests.get(BACKUP_URL, params=params, timeout=10)
                data = response.json()

                # If still error, propagate exception
                if "Error" in data:
                    raise RuntimeError(f"AGI PlaceSearch backup error: {data['Error']}")
            else:
                # Trial mode should not fallback; error is terminal
                return data

        # Success: return parsed JSON data
        return data

    except requests.RequestException as req_exc:
        # Network or HTTP-level error occurred
        if is_live:
            try:
                # Fallback to backup URL on network failure
                response = requests.get(BACKUP_URL, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()
                return data
            except Exception as fallback_exc:
                raise RuntimeError("AGI PlaceSearch unreachable on both endpoints") from fallback_exc

        else:
            # In trial mode, propagate the network exception
            raise RuntimeError(f"AGI trial error: {str(req_exc)}") from req_exc

File: address-geocode-international-nodejs/REST/place_search_rest.js
import axios from 'axios';
import querystring from 'querystring';
import { PSResponse } from './agi_response.js';

/**
 * @constant
 * @type {string}
 * @description The base URL for the live ServiceObjects Address Geocode International (AGI) API service.
 */
const LiveBaseUrl = 'https://sws.serviceobjects.com/agi/api.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the backup ServiceObjects Address Geocode International (AGI) API service.
 */
const BackupBaseUrl = 'https://swsbackup.serviceobjects.com/agi/api.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the trial ServiceObjects Address Geocode International (AGI) API service.
 */
const TrialBaseUrl = 'https://trial.serviceobjects.com/agi/api.svc/json/';

/**
 * <summary>
 * Checks if a response from the API is valid by verifying that it either has no Error object
 * or the Error.TypeCode is not equal to '3'.
 * </summary>
 * <param name="response" type="Object">The API response object to validate.</param>
 * <returns type="boolean">True if the response is valid, false otherwise.</returns>
 */
const isValid = (response) => !response?.Error || response.Error.TypeCode !== '3';

/**
 * <summary>
 * Constructs a full URL for the PlaceSearch API endpoint by combining the base URL
 * with query parameters derived from the input parameters.
 * </summary>
 * <param name="params" type="Object">An object containing all the input parameters.</param>
 * <param name="baseUrl" type="string">The base URL for the API service (live, backup, or trial).</param>
 * <returns type="string">The constructed URL with query parameters.</returns>
 */
const buildUrl = (params, baseUrl) =>
    `${baseUrl}PlaceSearch?${querystring.stringify(params)}`;

/**
 * <summary>
 * Performs an HTTP GET request to the specified URL with a given timeout.
 * </summary>
 * <param name="url" type="string">The URL to send the GET request to.</param>
 * <param name="timeoutSeconds" type="number">The timeout duration in seconds for the request.</param>
 * <returns type="Promise<PSResponse>">A promise that resolves to a PSResponse object containing the API response data.</returns>
 * <exception cref="Error">Thrown if the HTTP request fails, with a message detailing the error.</exception>
 */
const httpGet = async (url, timeoutSeconds) => {
    try {
        const response = await axios.get(url, { timeout: timeoutSeconds * 1000 });
        return new PSResponse(response.data);
    } catch (error) {
        throw new Error(`HTTP request failed: ${error.message}`);
    }
};

/**
 * <summary>
 * Provides functionality to call the ServiceObjects Address Geocode International (AGI) API's PlaceSearch endpoint,
 * retrieving geocoded location information for a given address or place with fallback to a backup endpoint for reliability in live mode.
 * </summary>
 */
const PlaceSearchClient = {
    /**
     * <summary>
     * Asynchronously invokes the PlaceSearch API endpoint, attempting the primary endpoint
     * first and falling back to the backup if the response is invalid (Error.TypeCode == '3') in live mode.
     * </summary>
     * @param {string} SingleLine - The full address on one line. Optional; for best results, use parsed inputs.
     * @param {string} Address1 - Address Line 1 of the international address. Optional.
     * @param {string} Address2 - Address Line 2 of the international address. Optional.
     * @param {string} Address3 - Address Line 3 of the international address. Optional.
     * @param {string} Address4 - Address Line 4 of the international address. Optional.
     * @param {string} Address5 - Address Line 5 of the international address. Optional.
     * @param {string} Locality - The name of the locality (e.g., city, town). Optional.
     * @param {string} AdministrativeArea - The administrative area (e.g., state, province). Optional.
     * @param {string} PostalCode - The postal code. Optional.
     * @param {string} Country - The country name or ISO 3166-1 Alpha-2/Alpha-3 code. Required.
     * @param {string} Boundaries - A comma-delimited list of coordinates for search boundaries. Optional; not currently used.
     * @param {string} MaxResults - Maximum number of results (1-10). Defaults to '10'.
     * @param {string} SearchType - Type of search (e.g., 'BestMatch', 'All'). Defaults to 'BestMatch'.
     * @param {string} Extras - Comma-delimited list of extra features. Optional.
     * @param {string} LicenseKey - Your license key to use the service. Required.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers. Defaults to true.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service. Defaults to 15.
     * @returns {Promise<PSResponse>} - A promise that resolves to a PSResponse object.
     */
    async invokeAsync(SingleLine, Address1, Address2, Address3, Address4, Address5, Locality, AdministrativeArea, PostalCode, Country, Boundaries, MaxResults = '10', SearchType = 'BestMatch', Extras, LicenseKey, isLive = true, timeoutSeconds = 15) {
        const params = {
            SingleLine,
            Address1,
            Address2,
            Address3,
            Address4,
            Address5,
            Locality,
            AdministrativeArea,
            PostalCode,
            Country,
            Boundaries,
            MaxResults,
            SearchType,
            Extras,
            LicenseKey
        };

        // Remove null/undefined params to avoid empty query params
        //Object.keys(params).forEach(key => params[key] == null && delete params[key]);

        const url = buildUrl(params, isLive ? LiveBaseUrl : TrialBaseUrl);
        let response = await httpGet(url, timeoutSeconds);

        if (isLive && !isValid(response)) {
            const fallbackUrl = buildUrl(params, BackupBaseUrl);
            const fallbackResponse = await httpGet(fallbackUrl, timeoutSeconds);
            return fallbackResponse;
        }
        return response;
    },

    /**
     * <summary>
     * Synchronously invokes the PlaceSearch API endpoint by wrapping the async call
     * and awaiting its result immediately.
     * </summary>
     * @param {string} SingleLine - The full address on one line. Optional; for best results, use parsed inputs.
     * @param {string} Address1 - Address Line 1 of the international address. Optional.
     * @param {string} Address2 - Address Line 2 of the international address. Optional.
     * @param {string} Address3 - Address Line 3 of the international address. Optional.
     * @param {string} Address4 - Address Line 4 of the international address. Optional.
     * @param {string} Address5 - Address Line 5 of the international address. Optional.
     * @param {string} Locality - The name of the locality (e.g., city, town). Optional.
     * @param {string} AdministrativeArea - The administrative area (e.g., state, province). Optional.
     * @param {string} PostalCode - The postal code. Optional.
     * @param {string} Country - The country name or ISO 3166-1 Alpha-2/Alpha-3 code. Required.
     * @param {string} Boundaries - A comma-delimited list of coordinates for search boundaries. Optional; not currently used.
     * @param {string} MaxResults - Maximum number of results (1-10). Defaults to '10'.
     * @param {string} SearchType - Type of search (e.g., 'BestMatch', 'All'). Defaults to 'BestMatch'.
     * @param {string} Extras - Comma-delimited list of extra features. Optional.
     * @param {string} LicenseKey - Your license key to use the service. Required.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers. Defaults to true.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service. Defaults to 15.
     * @returns {PSResponse} - A PSResponse object with geocoded location details or an error.
     */
    invoke(SingleLine, Address1, Address2, Address3, Address4, Address5, Locality, AdministrativeArea, PostalCode, Country, Boundaries, MaxResults = '10', SearchType = 'BestMatch', Extras, LicenseKey, isLive = true, timeoutSeconds = 15) {
        return (async () => await this.invokeAsync(
            SingleLine, Address1, Address2, Address3, Address4, Address5, Locality, AdministrativeArea, PostalCode, Country, Boundaries, MaxResults, SearchType, Extras, LicenseKey, isLive, timeoutSeconds
        ))();
    }
};

export { PlaceSearchClient, PSResponse };